### PR TITLE
Cycle through all Armament.LocalOffsets if weapon has Burst: 1

### LINKED ISSUE #14464: The APC's Armament LocalOffset is only utilizing one of its values
In Tiberium Dawn the APC's Armament LocalOffset doesn't currently alternate between the two values (it only shoots out of one of its two barrels), this is because the APC weapon doesn't use bursts. Giving the APCGun a burst will fix this.

### LINKED ISSUE #14577: Longbow Armament LocalOffset is only utilizing one of its values
In Red Alert the Longbow's Armament LocalOffset doesn't currently alternate between left and right (it only shoots out of one of its two wings). Similar to #14464

### PR DESCRIPTION
**Pull request by reaperrr**
While the approach in #14578 might be good enough in some cases, the fact that modders are forced to use BurstDelay < ReloadDelay and therefore couldn't balance this 100% perfectly just highlights that this needs a code-level solution.

Also, TS was still using values from a time when BurstDelay == ReloadDelay was supposed to be supported (but dropped due to other technical issues that would have caused).

Fixes #14464.
Fixes #14577.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 1]
```diff
@@ -232,7 +236,10 @@ public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
 
 			ticksSinceLastShot = 0;
 
-			var barrel = Barrels[Burst % Barrels.Length];
+			// If Weapon.Burst == 1, cycle through all LocalOffsets, otherwise use the offset corresponding to current Burst
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Armament.cs]: Can you please document this in the `LocalOffset` desc too?

**Review by fruestueck — COMMENTED** [review]
Tested TD with APC and worked fine. Can't tell much to the code.

**Comment by reaperrr** [comment]
Updated. Fixes  #14577 as well now.

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Armament.cs]: Done.

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 2]
```diff
@@ -39,7 +39,8 @@ public class ArmamentInfo : PausableConditionalTraitInfo, Requires<AttackBaseInf
 		[Desc("Time (in frames) until the weapon can fire again.")]
 		public readonly int FireDelay = 0;
 
-		[Desc("Muzzle position relative to turret or body. (forward, right, up) triples")]
+		[Desc("Muzzle position relative to turret or body, (forward, right, up) triples.",
+			"If weapon Burst = 1, it cycles through all listed offsets, otherwise the offset corresponding to current burst is used.")]
```
Comment (chrisforbes) [inline OpenRA.Mods.Common/Traits/Armament.cs]: current shot in the burst, right?

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 3]
```diff
@@ -232,7 +237,10 @@ public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
 
 			ticksSinceLastShot = 0;
 
-			var barrel = Barrels[Burst % Barrels.Length];
+			// If Weapon.Burst == 1, cycle through all LocalOffsets, otherwise use the offset corresponding to current Burst
+			currentBarrel = currentBarrel <= barrelCount - 1 ? currentBarrel : 0;
```
Comment (chrisforbes) [inline OpenRA.Mods.Common/Traits/Armament.cs]: `currentBarrel %= barrelCount;` ?

**Review by chrisforbes — DISMISSED** [review]

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 2 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/Armament.cs]: According to what I found online, 'burst' stands for a shot in a salvo.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Armament.cs]: See #13090 in relation to this ongoing confusion.

**Comment by reaperrr** [comment]
Updated.

**Comment by AoAGeneral** [comment]
With this change in mind does this affect Apache and Orcas as well? When this also gets implemented in does this create extra damage for the APC?

**Comment by GraionDilach** [comment]
This change is plainly visual.

**Comment by GraionDilach** [comment]
Could this logic be expanded to apply to all cases where the amount of the firing offsets is a direct multiplication of burst? (say, Burst: 2 with 4 offsets).

**Comment by reaperrr** [comment]
> With this change in mind does this affect Apache and Orcas as well?

Yes.

> When this also gets implemented in does this create extra damage for the APC?

No.
This change is 100% cosmetical, it has zero impact on range, targeting, dps or anything.
Basically, in all RA/TD cases this will work just as
```
ReloadDelay: 20
Burst: 2
BurstDelays: 20
```
would, if BurstDelays were allowed to be equal to ReloadDelay (which they're not for technical reasons).

**Comment by reaperrr** [comment]
> Could this logic be expanded to apply to all cases where the amount of the firing offsets is a direct multiplication of burst? (say, Burst: 2 with 4 offsets).

I don't feel like coming up with the code and math for that right now, but generally speaking I don't see why not.

**Review by pchote — DISMISSED** [review]
LGTM and confirmed the noted fixes ingame.

However, this doesn't fix the TD Orca, despite that having two firing offsets defined?

**Comment by pchote** [comment]
The Orca uses the same kind of workaround as TS, which should now be removed.

**Comment by reaperrr** [comment]
Fixed.

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/63f5720b44c6043c46e96e1f3190537761c426de

**Pull request merged**